accepted/00052-new-package-predownloading.md

```diff
@@ -0,0 +1,86 @@
+  - package installation: https://server.tf.local/rhn/systems/details/packages/InstallPackages.do?sid=1000010000&
+  - package upgrade: https://server.tf.local/rhn/systems/details/packages/UpgradableList.do?sid=1000010000&
+
+- SSM:
+  - patch application: https://server.tf.local/rhn/systems/ssm/ListErrata.do
+  - package installation: https://server.tf.local/rhn/ssm/PackageInstall.do
+  - package upgrade: https://server.tf.local/rhn/ssm/PackageUpgrade.do
+
+*Links above assume `server.tf.local` as the server location and `1000010000` as a minion ID.*
+
+### API
+
+The following calls will need a new optional `downloadOnly` boolean flag:
+- `actionchain.addErrataUpdate`
+- `actionchain.addPackageInstall`
+- `actionchain.addPackageUpgrade`
+- `actionchain.schedulePackageInstall`
+- `actionchain.schedulePackageInstallByNevra`
+- `system.schedulePackageInstall`
+- `system.schedulePackageInstallByNevra`
+- `system.scheduleApplyErrata`
+- `systemgroup.scheduleApplyErrataToActive`
+
+## Implementation
+- new checkbox item will be added to current JSPs/React components
+  - checkbox has to be shown only if applicable systems are all minions
+- existing Java handling code will receive the new flag, pass it down to Salt calling code in order to apply
+  - [pkgdownload.sls](https://github.com/uyuni-project/uyuni/blob/c8ffe6b9425392f5235864ad070646bb8ebc2ecb/susemanager-utils/susemanager-sls/salt/packages/pkgdownload.sls) instead of [pkginstall.sls](https://github.com/uyuni-project/uyuni/blob/c8ffe6b9425392f5235864ad070646bb8ebc2ecb/susemanager-utils/susemanager-sls/salt/packages/pkginstall.sls) if set
+  - [patchdownload.sls](https://github.com/uyuni-project/uyuni/blob/c8ffe6b9425392f5235864ad070646bb8ebc2ecb/susemanager-utils/susemanager-sls/salt/packages/patchdownload.sls) instead of [patchinstall.sls](https://github.com/uyuni-project/uyuni/blob/c8ffe6b9425392f5235864ad070646bb8ebc2ecb/susemanager-utils/susemanager-sls/salt/packages/patchinstall.sls) if set
+  - alternatively, those files might be unified and use `pkg.downloaded` instead of `pkg.installed` conditionally (resp. `pkg.patch_downloaded` instead of `pkg.patch_installed`)
+- existing configuration options are to be deprecated (in SUSE Manager supported versions) or removed
+  - existing code handling the configuration options can be removed from Uyuni
+
+# Drawbacks
+ - this is limited to Salt minions. Traditional clients retain their existing mechanism and are not affected
+ - as the `pkg.downloaded` state [is not yet supported in Salt for Debian systems](https://docs.saltstack.com/en/2018.3/ref/states/all/salt.states.pkg.html#salt.states.pkg.downloaded), this remains an exclusive for `yum` and `zypper` (this RFC does not change this aspect)
+ - current users of the existing feature will have to adapt their workflows
+ - proposed mechanism is conceptually different from the one in place for traditional clients
```

Comment:
On big drawback was not named here, but in the alternative approach:

- with the UI, users have to go through the package/patch selection twice

The big problem I see is, that when a user made a very individual selection he is nearly unable to reproduce it a second time for the real action execution.

The result might be, that he choose a different selection then for the download and additional packages needs to be downloaded when the real action will be executed.

Reply:
@moio - I think you suggestion in the comments about using a text field instead of a checkbox where a user can enter the number of hours a download should happen before the real action is executed is a good idea.

We hear already complains for the SP Migration UI where we have "Dry Run" vs. "Real execution" and there are the options for spection selections much smaller then here.

I think we should take the extra effort.